enhancements/installer/azure-internal-with-user-defined-routing.md

```diff
@@ -0,0 +1,243 @@
+---
+title: azure-internal-with-user-defined-outbound-routing
+authors:
+  - "@abhinavdahiya"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-20
+last-updated: 2020-05-20
+status: implementable
+see-also:
+  - "./enhancements/azure-private-internal-clusters.md"
+  - https://docs.microsoft.com/en-us/azure/aks/egress-outboundtype
+  - https://github.com/openshift/installer/pull/3324
+replaces:
+superseded-by:
+---
+
+# Azure Internal Clusters with User Defined Outbound Routing
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently for internal clusters on Azure, OpenShift clusters always use Public Standard Loadbalancers for Internet egress. This configuration requires public IPs, public Loadbalancers etc. which some customers do not want for their internal clusters. This proposal allows the users to choose their own outbound routing for Internet allowing them to use any pre-existing setups instead of the per-cluster OpenShift recommended way.
+
+## Motivation
+
+Internal clusters on Azure require explicit setup to allow outbound routing or egress to Internet to pull container images, updates, access cloud APIs etc. Currently the cluster uses a public standard loadbalancer, with only outbound_rules, to provide the outbound routing to Internet. But this force customers that either have pre-existing setups like PROXY for outbound routing or do not require any egress to Internet at all i.e. air-gapped, into the above presrcribed setup without allowing them to change it. And often for customers that are interested in internal clsuters already have policies that disallow creation resources like public IP addresses, which causes the installations to fail and forces the customers to use user-provided workflow.
+
+Allowing the customer to take over the responsibility of routing allows the user flexibility of using the installer-provisioned workflow.
+
+### Goals
+
+1. Allow users to specify UserDefinedRouting for outbound with pre-exiting networks.
+
+### Non-Goals
+
+## Proposal
+
+### Configuration
+
+The install-config.yaml should include a new field for Azure platform that allows the users to configure the outbound type to Loadbalancer or UserDefined, with the default being Loadbalancer
+
+The user should only be allowed to change the outbound type when using pre-exiting networking as outbound routing needs to be setup by user before installing the cluster into the network, and only for Internal publishing strategy only as for External clusters the load balancers already exist to provide easy egress.
```

Comment:
+1 on outbound type depending on BYO-vnet.

However, being dependent on internal publishing strategy seems overly restrictive.  I am confident that there will be customers who want the external publishing strategy and cluster traffic egress via their networks, and I don't think there's a technical blocker to allowing this architecture.